org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/StringSwitchJavacFilter.java

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.LabelNode;
+import org.objectweb.asm.tree.LookupSwitchInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TableSwitchInsnNode;
+
+/**
+ * Filters code that is generated by javac for a switch statement with a String.
+ */
+public final class StringSwitchJavacFilter implements IFilter {
+
+	public void filter(final String className, final String superClassName,
+			final MethodNode methodNode, final IFilterOutput output) {
+		AbstractInsnNode i = methodNode.instructions.getFirst();
+		while (i != null) {
+			filter(i, output);
+			i = i.getNext();
+		}
+	}
+
+	/**
+	 * javac generates two switches. First one by {@link String#hashCode()}.
+	 * Number of handlers in the second switch is equal to number of handlers in
+	 * source code, so it is enough to completely filter-out first switch.
+	 * Handler for default case of the first switch - is the second switch.
+	 */
+	private void filter(final AbstractInsnNode start,
+			final IFilterOutput output) {
+		final LabelNode dflt;
+		if (start.getOpcode() == Opcodes.LOOKUPSWITCH) {
+			dflt = ((LookupSwitchInsnNode) start).dflt;
+		} else if (start.getOpcode() == Opcodes.TABLESWITCH) {
+			dflt = ((TableSwitchInsnNode) start).dflt;
+		} else {
+			return;
+		}
+		if (new Matcher().match(start, dflt)) {
+			output.ignore(start, dflt);
+		}
+	}
+
+	private static class Matcher extends AbstractMatcher {
+		boolean match(final AbstractInsnNode start,
+				final AbstractInsnNode secondSwitchLabel) {
+			cursor = start;
+			for (int i = 0; cursor != null && i < 4; i++) {
+				cursor = cursor.getPrevious();
+			}
+			if (cursor == null || cursor.getOpcode() != Opcodes.ICONST_M1) {
```

Comment:
Couldn't that be `nextIs(Opcodes.ICONST_M1)` followed by `if (cursor == null) return;`?

Reply:
@marchof Unfortunately no: `nextIs` first moves to next instruction and then checks, so can't be used to verify current instruction, and we can't place cursor before very first instruction of method.

Reply:
@marchof this is kind of catch 22 - either we need to special case checking of beginning, or will have problem with end of methods, i.e. if we change semantic and cursor will be pointing to instruction for checking and moved after checking, then `null` can not be used to represent no-match as it is now, because `next` of last instruction is `null`. Maybe we can represent no-match by a special non-null value, but will be better to do this separately from this PR as it requires change of semantics `next` methods.

Reply:
@marchof another option is to have fake instruction at the beginning:
```
private void moveBack(int num) {
	for (int i = 0; i < num; i++) {
		if (cursor.getPrevious() == null) {
			cursor = new InsnNode(Opcodes.NOP) {
				final AbstractInsnNode next = cursor;
				@Override
				public AbstractInsnNode getNext() {
					return next;
				}
			};
		} else {
			cursor = cursor.getPrevious();
		}
	}
}
```

But I doubt that this is a good idea in general and definitely too much of complexity if done only here, compared to current handling of beginning.